Add ability to store/load surrogate data and improve basic surrogate model API

aeslaughter opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Re-factor mesh meta data as generic 'extra' data that apps could create

(refs #14875)

aeslaughter committed:
Add/update tests for restartable meta data changes

(refs #14875)

aeslaughter committed:
Improve assert so that an empty if doesn't appear

(refs #14875)

aeslaughter committed:
Implement save/load of surrogate model data

(refs #14875)

aeslaughter committed:
Update documentation

(refs #14875)

aeslaughter committed:
Re-factor and Fixes after rebase

(refs #14875)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/486079/) on 414e3fc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14935/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14935/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6423990678dfc3a14c73bb4794422d9a0e09a1be`

aeslaughter commented:
@zachmprince I fiddled around with your code a bit to make things a bit more consistent. 
- I changed the PolynomialChaosSobolStats class to use initialSetup(). The SurrorgateModel and this VPP are both GeneralUserObjects, so the setup and execute dependencies are handled
- I also removed threadJoin from these classes, since GeneralUserObjects are not threaded

zachmprince commented on modules/stochastic_tools/src/utils/PolynomialQuadrature.C:
```diff
@@ -31,20 +32,27 @@ makePolynomial(const Distribution * dist)
   const BoostNormalDistribution * bn_dist = dynamic_cast<const BoostNormalDistribution *>(dist);
   if (bn_dist)
-    return libmesh_make_unique<const Hermite>(bn_dist);
+    return libmesh_make_unique<Hermite>(bn_dist);
```
Why remove the `const`? All the public member functions are `const` anyway.

aeslaughter replied:
Mistake. One of my iterations to the dataLoad/Store required this, but I came up with a better way.

zachmprince reviewed (changes requested):
Man, this is looking really nice. I'll put WIP in #14934 since there will be conflicts and this is more important to merge.

aeslaughter commented on modules/stochastic_tools/src/surrogates/PolynomialChaos.C:
```diff
@@ -20,46 +20,79 @@ PolynomialChaos::validParams()
+
+  // If training data is loaded, it is store in a single file so the data must be partitioned
+  if (!isTraining() && n_processors() > 1)
+    _coeff.assign(_coeff.begin() + _local_coeff_begin, _coeff.begin() + _local_coeff_end);
```
No, I just dropped this in at the last minute. I need to come up with an API for outputting as well. I think if you were to run in parallel now then only the part from rank 0 would show up in the file.

I think we will need to add "parallel_type" to SurrogateModel objects as well. And the output will need to be aware and do the correct thing. The SurrogateModel will need to be able to run in parallel and write out a single file or one file per process. Also, when loading it should be able to read a single file and setup a distributed model or read multiple files into a distributed run. We also need to thing about training on 100 cores, but evaluating on 10.

Initially, I want to just focus on a single file load/store, but handle the serialization and de-serialization of the model data. For this the SurrogateModel will probably need an "onLoad" and "onStore" methods to handle the data. I am still thinking about the API for this...it get complicated because if you put all the data on rank 0 for output but keep going in the solve/training now the data is not distributed, etc.

I probably need to come up with some test cases so I can understand what needs to be supported. Perhaps we pick the most common case and error on the others first.

zachmprince replied:
Wow, parallel stuff is hard. Would it help if I came up with a way to distribute the coefficients? I think this would require communication on each call to `evaluate`. Perhaps we could add `evaluateInitialize` and `evaluateFinalize` to take care of this. I have a feeling this will be useful for other surrogate models.

aeslaughter replied:
Like I said on the call. I thought it was distributed, so this change is just wrong. For now lets assume that all training data is "replicated". This will make things a lot easier as we ramp up.

I think I will add "parallel_type" put only have "replicated" as an option. I will also add a parallel test where it is trained with 3 cores and executed with 2.

aeslaughter commented on modules/stochastic_tools/src/surrogates/PolynomialChaos.C:
```diff
@@ -20,46 +20,79 @@ PolynomialChaos::validParams()
-    _coeff(_ncoeff)
+    _ndim(declareModelData<unsigned int>("_ndim")),
+    _tuple(declareModelData<std::vector<std::vector<unsigned int>>>("_tuple")),
+    _ncoeff(declareModelData<std::size_t>("_ncoeff")),
```
Cool, didn't think about that.

aeslaughter commented on modules/stochastic_tools/include/surrogates/SurrogateModel.h:
```diff
@@ -0,0 +1,103 @@
+  virtual void initialize() final;
+  virtual void execute() final;
+  virtual void finalize() final;
+  virtual void threadJoin(const UserObject &) final {} // GeneralUserObjects are not threaded
```
It should be and I put it there years ago. But, someone came along and created ThreadedGeneralUserObject, which I think inherits from this, so had to take it out.

I thought about tackling that, but then decided I don't care anymore since I already did it once.

zachmprince replied:
How difficult would it be to have threading for our surrogate models? The extra level of parallelization could be beneficial.

aeslaughter replied:
It is quite difficult and probably not worth the effort. The main items that are threaded in MOOSE are the element and node loops, which are well defined. What type of work and data and how it is used is not well defined for surrogates, so generalizing that case would be difficult. Furthermore, threading code is way more difficult than MPI code.

aeslaughter commented on modules/stochastic_tools/include/utils/PolynomialQuadrature.h:
```diff
@@ -87,28 +87,29 @@ class Legendre : public Polynomial
 {
 public:
+  Hermite(const Real mu, const Real sig);
   Hermite(const NormalDistribution * dist);
```
@zachmprince I am thinking that we don't need these constructors. The make_polynomial function can take a distribution and then just call the constructor I created above. What to you think?

zachmprince replied:
That could be a good idea. Do you think that would look cleaner? We would still need to dynamic cast and gather input parameters.

aeslaughter replied:
Yes, the makePolynomial would be about the same. But, the Polynomial objects would be cleaner. They are simply selected based on the distribution but do not have a direct dependence. I will add this to the PR.

aeslaughter commented:
@zachmprince Thanks for the comments, I think the design is starting to take shape. Next week 
I will clean up things based on the comments here and fix the segfault this happening all over the place.

Once it is all green I will play around with the idea of a separate Trainer object that informs the a Model, just to see where it goes.

aeslaughter committed:
Address reviewer comments

(refs #14875)

aeslaughter committed:
Improve logic and testing for PolynomialChaos for replicated and distributed training data

(refs #14875)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/487802/) on 57de4d3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14935/site)

This comment will be updated on new commits.

zachmprince commented on modules/stochastic_tools/src/surrogates/PolynomialChaos.C:
```diff
@@ -78,7 +111,7 @@ PolynomialChaos::execute()
     for (std::size_t i = 0; i < _ncoeff; ++i)
     {
-      Real val = _values[p - _sampler->getLocalRowBegin()];
+      Real val = (*_values_ptr)[p - _sampler->getLocalRowBegin()];
```
I think this causing the bug. If you switch to below, it should work
```suggestion
      Real val = (*_values_ptr)[p];
```

aeslaughter replied:
I will look into that. What doesn't make sense is that the current version, which is also working has this.

https://github.com/idaholab/moose/blob/72914c17e0622b943c2df646e6bb2ee69824bd1d/modules/stochastic_tools/src/userobject/PolynomialChaos.C#L81

aeslaughter replied:
I would expect that this index would need to change based how how the _values data setup in parallel: replicated (don't need to substract) and distributed (do need to subtract). I was playing around with StochasticResults and making sure it was properly replicated, which it wasn't. But, that wasn't the original PR.

zachmprince replied:
Correct, we need to shift the values if distributed. I'm not sure the best way to access the parallel type of the VPP though.

aeslaughter replied:
There is no method now, I will add it in along with a test that uses distributed data. Thanks again for helping me figure this one out.

zachmprince replied:
No problem. Glad I could help. Let me know if you run into any more problems.

zachmprince commented on modules/stochastic_tools/test/tests/surrogates/poly_chaos/master_2d_mc.i:
```diff
@@ -80,6 +80,7 @@
 
 [Surrogates]
   [poly_chaos]
+    #type = PolynomialChaosOld
```
Left over from debugging?

aeslaughter replied:
Dang.

zachmprince commented on modules/stochastic_tools/test/tests/surrogates/poly_chaos/gold/master_2d_mc_out_pc_coeff_0002.csv:
```diff
@@ -0,0 +1,16 @@
+0.067633779358418,3,1
+-0.044360872953855,2,2
+-0.17515912643907,1,3
+-0.058910304264404,0,4
```
Did these values change, or were they copied over?

aeslaughter replied:
Copies

zachmprince commented on modules/stochastic_tools/test/tests/surrogates/load_store/tests:
```diff
@@ -10,15 +10,57 @@
+      allow_warnings = true
+      max_parallel = 2
+
+      detail = "evaluated separately with a different number of processors."
```
So it works with different number of processors?

aeslaughter replied:
Yes, and with replicated or distributed data.

zachmprince reviewed (commented):
Looks great! Just a few questions.

aeslaughter commented:
I would like this merged. I will then start on separating the training and evaluating into two different objects.

permcody commented:
Looks like there's no rush with the build systems down.

aeslaughter committed:
remove duplicates created in rebase

(refs #14875)

aeslaughter committed:
Move dynamic_cast from calls to the TheWarehouse

For some reason the dynamic_cast performed within the TheWarehouse fails for the combined module if the
container supplied to queryInto is std::vector<SurrogateModel*>

(refs #14875)

aeslaughter commented on modules/stochastic_tools/src/actions/LoadSurrogateDataAction.C:
```diff
@@ -29,15 +29,18 @@ LoadSurrogateDataAction::LoadSurrogateDataAction(InputParameters params) : Actio
-  std::vector<SurrogateModel *> models;
-  _app.theWarehouse().query().condition<AttribSystem>("UserObject").queryInto(models);
-  for (auto model_ptr : models)
-    if (model_ptr->isParamValid("filename"))
```
@rwcarlsen queryIntro doesn't seem to handle the dynamic_cast correctly. If I do what is here then the combined modules app fails the cast inside of TheWarehouse::queryInto

aeslaughter commented:
@permcody This is ready to go. Thanks.

permcody commented:
acknowledged

aeslaughter commented on modules/stochastic_tools/src/surrogates/PolynomialChaos.C:
```diff
@@ -20,51 +21,93 @@ PolynomialChaos::validParams()
+  if (isTraining())
+  {
+    // Setup data needed for training
+    _order = getParam<unsigned int>("order");
```
_order is not set if isTraining is false, but all the tests pass. How is this possible?

permcody replied:
Getting lucky? Turn on Valgrind testing and see if this fails with an uninitialized error.

aeslaughter replied:
We figure it is initialized to something large. The way it is used would just result in a huge amount of data being unused. I will have a fix very soon.

aeslaughter committed:
Correctly initialize _order

(refs #14875)

permcody commented:
Feel free to clean up you bugs with a squash occasionally. No need to have a known bug in your history. 

```
git fixup
git rebase --autosquash
```

aeslaughter commented:
@permcody I do when I know which commit it belongs; this has been around a while and has two significant rebases along the way that were full of conflicts. This also has a lot of comments on parts of the code that I don't want to blow away.

Anyway I am closing this in favor of #15030 which builds on this. I will clean that up a bit.
